Chore: Rewrite `ConfirmButton`

ashharrison90 opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

- rewrites `ConfirmButton` as a functional component
- fixes some a11y issues (focus now properly maintains between initial button/confirm button)
  - removes the `autoFocus` prop since it should just be the default behaviour
  - removes any uses of `autoFocus` in the grafana codebase
- adjusts the positioning of the confirm/cancel buttons to use a `fixed` strategy to not cause page overflowing
  - you can observe this problem in storybook in main
  - set your scrollbars to appear always (settings -> appearance)
  - go to the `ConfirmButton` story
  - note the horizontal scrollbar at the bottom of the page
  - repeat the same steps but on this branch
  - no scrollbar! 🥳 

**Why do we need this feature?**

- see above

**Who is this feature for?**

- everyone

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

ashharrison90 committed:
convert to function component, clean up + fix a11y

ashharrison90 committed:
use position: fixed;

eledobleefe commented:
Just curious, are we moving towards turning components into functional ones? 🤔

ashharrison90 commented:
> Just curious, are we moving towards turning components into functional ones? 🤔

my 2c: 
- for new components: any new component we write should be functional
- for existing components:
  - if they're easy/small enough (like this one) and you're planning on refactoring the component in some way, then just convert it.
  - if they're big legacy components we should probably just leave them alone and let them naturally be replaced by newer functional components.

ashharrison90 commented on packages/grafana-ui/src/components/ConfirmButton/ConfirmButton.tsx:
```diff
@@ -24,145 +24,109 @@ export interface Props extends Themeable2 {
   confirmVariant?: ButtonVariant;
```
it's used here: https://github.com/grafana/grafana/pull/84402/files#diff-9ce9a2afd3198c5a910502b703a19b9a26f52d456e9b67d95327c799d9490273R110

eledobleefe commented:
I saw this when running on local, but I also checked in the main branch and it still happens. I'll raise an issue.
https://github.com/grafana/grafana/assets/65417731/805276e9-ca45-41e1-8692-738fc145650c
Apart from this, LGTM 😅 And sorry again for the misunderstood.

ashharrison90 commented:
> I saw this when running on local, but I also checked in the main branch and it still happens. I'll raise an issue. https://github.com/grafana/grafana/assets/65417731/805276e9-ca45-41e1-8692-738fc145650c Apart from this, LGTM 😅 And sorry again for the misunderstood.

@eledobleefe ah nice, that's my fault. i'll fix in this PR, don't bother raising an issue 👍

eledobleefe commented:
> > I saw this when running on local, but I also checked in the main branch and it still happens. I'll raise an issue. https://github.com/grafana/grafana/assets/65417731/805276e9-ca45-41e1-8692-738fc145650c Apart from this, LGTM 😅 And sorry again for the misunderstood.
> 
> @eledobleefe ah nice, that's my fault. i'll fix in this PR, don't bother raising an issue 👍

Why is it your fault? I though it was missing `position:relative` in the row 🤔

ashharrison90 committed:
fix

ashharrison90 commented:
@eledobleefe naw it's because i switched to `position: fixed;`

i've updated the logic. there is one _slight_ downside in that it will show a scrollbar whilst the animation is happening, but 🤷  it's better than before

eledobleefe commented:
> @eledobleefe naw it's because i switched to `position: fixed;`
> 
> i've updated the logic. there is one _slight_ downside in that it will show a scrollbar whilst the animation is happening, but 🤷 it's better than before

Yeah, it's weird that a scrollbar appears in each axis 😅 And if it was you, why was it failing in main too?

ashharrison90 commented:
> > @eledobleefe naw it's because i switched to `position: fixed;`
> > i've updated the logic. there is one _slight_ downside in that it will show a scrollbar whilst the animation is happening, but 🤷 it's better than before
> 
> Yeah, it's weird that a scrollbar appears in each axis 😅 And if it was you, why was it failing in main too?

i don't think it was for me on main, sooooo.. pass 😅
